data_pipeline_api/file_formats/object_file.py

```diff
@@ -0,0 +1,55 @@
+from io import IOBase
+import h5py
+import pandas as pd
+import numpy as np
+from functools import reduce
+from operator import getitem
+from typing import Union, List
+
+
+Table = pd.DataFrame
+Array = np.ndarray
+
+
+def get_components(file: IOBase) -> List[str]:
+    components = []
+
+    def add_dataset_parent(name, obj):
+        if isinstance(obj, h5py.Dataset):
+            components.append(obj.parent.name)
+
+    h5py.File(file, mode="r").visititems(add_dataset_parent)
+    return components
+
+
+def get_read_group(file: IOBase, component: str) -> h5py.Group:
+    return h5py.File(file, mode="r").get(component)
+
+
+def get_write_group(file: IOBase, component: str) -> h5py.Group:
+    return h5py.File(file, mode="a").require_group(component)
+
+
+def read_table(file: IOBase, component: str) -> Table:
+    return pd.DataFrame(get_read_group(file, component)["table"][()]).apply(
+        lambda s: s.str.decode("utf-8")
+        if pd.api.types.is_object_dtype(s.infer_objects())
+        else s
+    )
+
+
+def write_table(file: IOBase, component: str, table: Table):
+    records = table.to_records(index=False)
```

Comment:
Why are you not using df.to_hdf here?

Reply:
I don't believe it will create the correct internal structure, and so the resulting file wouldn't conform to the API.

Reply:
If this is not the case, or if you have other concerns about our internal hdf5 format, do let me know.

Reply:
I just wanted to understand the reason. So the problem is that `key` is actually the component and "table" is some other concept that the to_hdf function does not expose?

Reply:
Yes, table is the structure defined in the "standard API" spec. It isn't just that table is a concept that `to_hdf` doesn't expose, but that `to_hdf` imposes some structure of its own that is orthogonal to that required by table (afaik).

Reply:
There is some information about it in the [user guide](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#table-format).